torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -120,44 +123,161 @@ def can_auto_functionalize(op: OperatorBase) -> bool:
+            serialize_single_view(
+                f"_{arg_name}",
+                input_kwargs[arg_name],
+                arg_to_base_index.get(arg_name, None),
+            )
+        else:
+            raise RuntimeError(f"Unsupported type {arg_type}")
+
+
+def deserialize_views_meta(arg_names, arg_types, input_kwargs, all_bases, pop_args):
+    def get_arg(name):
+        if pop_args:
+            return input_kwargs.pop(name)
+
+        return input_kwargs[name]
+
+    def deserialize_single_view(prefix):
+        base_index = get_arg(f"{prefix}_base_index")
+        if base_index is None:
+            return None
+        elif f"{prefix}_size" not in input_kwargs:
+            # This means the argument is an alias of the base
+            return ViewInfo(base_index, all_bases[base_index], is_alias_to_base=True)
+
+        else:
+            size = get_arg(f"{prefix}_size")
+            stride = get_arg(f"{prefix}_stride")
+            storage_offset = get_arg(f"{prefix}_storage_offset")
+            return ViewInfo(
+                base_index, size, stride, storage_offset, is_alias_to_base=False
+            )
+
+    args_view_info: Dict[str, Any] = {}
+    for arg_name, arg_type in zip(arg_names, arg_types):
+        if isinstance(arg_type, torch.ListType):
+            length = get_arg(f"_{arg_name}_length")
+            if length is None:
+                # The whole list is None.
+                args_view_info[arg_name] = None
+            else:
+                args_view_info[arg_name] = [
+                    deserialize_single_view(f"_{arg_name}_{i}") for i in range(length)
+                ]
+
+        elif isinstance(arg_type, (torch.TensorType, torch.OptionalType)):
+            args_view_info[arg_name] = deserialize_single_view(f"_{arg_name}")
+        else:
+            raise RuntimeError(f"Unsupported type {arg_type}")
+    return args_view_info
+
+
 @auto_functionalized.py_impl(DispatchKey.CompositeExplicitAutograd)
 def auto_functionalized_dense(
     _mutable_op: OpOverload,
-    _only_clone_these_tensors: Optional[Tuple[str, ...]] = None,
+    _only_clone_these_bases: Optional[Tuple[int, ...]] = None,
     **kwargs: Any,
 ) -> Tuple[Any, Tuple[Tensor, ...]]:
+    all_bases: List[Tensor] = kwargs.pop("_all_bases", [])
+    mutable_args_names, mutable_args_types = get_mutable_args(_mutable_op)
+    args_view_info = deserialize_views_meta(
+        mutable_args_names, mutable_args_types, kwargs, all_bases, pop_args=True
+    )
+
+    if _only_clone_these_bases is None:
+        _only_clone_these_bases = tuple(range(len(all_bases)))
+
+    def maybe_copy(i, t):
+        if t is None:
+            return None
+        if i in _only_clone_these_bases:
+            return t.clone()
+        else:
+            return t
+
+    all_bases_new = [maybe_copy(i, t) for i, t in enumerate(all_bases)]
+
+    def regenerate_view(ViewInfo):
+        if ViewInfo.is_alias_to_base:
+            return all_bases_new[ViewInfo.base_index]
```

Comment:
The word "alias" is a overloaded, it can mean "x is y" or "x = aten.alias(y)". In the second case, x is not y.

Can we name it `ViewInfo.is_view` instead?
Also, nit: regenerate_view can just be a method on the ViewInfo dataclass:
`this_base = view_info.regenerate_base(all_bases)`

Reply:
done